docs/design_proposals/artifact-dependencies.md

```diff
@@ -0,0 +1,361 @@
+
+## Builder interfaces
+There are two builder abstractions -- in [build.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/build.go#L37) and [parallel.go](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go#L34)
+
+```go
+type Builder interface {
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact) ([]Artifact, error)
+}
+```
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
+```
+The first describes a builder for a list of artifacts. There are a few implementations like `cache`, `local` and `cluster` builders.
+The second describes a per artifact builder. Again there are a few implementations like `docker`, `buildpacks`, etc.
+
+We modify both of them to:
+
+```go
+type Builder interface {
+	Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, existing []Artifact) ([]Artifact, error)
+}
+```
+```go
+type ArtifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string, artifactResolver ArtifactResolver) (string, error)
+```
+where we define `ArtifactResolver` interface, as:
+```go
+// ArtifactResolver provides an interface to resolve built artifacts by image name.
+type ArtifactResolver interface {
+	GetImageTag(imageName string) string
+}
+```
+
+This necessitates all multi-artifact builder implementations to also accept a slice of already built artifacts' information. All single artifact builders require an `ArtifactResolver` that can provide the required artifacts.
+
+## Build controller
+
+[InSequence](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/sequence.go) and [InParallel](https://github.com/GoogleContainerTools/skaffold/blob/10275c66a142719897894308b9e566953712a0fe/pkg/skaffold/build/parallel.go) are two build controllers for deciding how to schedule the run of multiple builds together. `InSequence` runs all builds sequentially whereas `InParallel` runs them parallely with a max concurrency defined by a `concurrency` field.
+
+After introducing inter-artifact dependencies we'll need to run the builds in a topologically sorted order.
+We introduce a new controller `scheduler.go` and deprecate `sequence.go` and `parallel.go`. Here we model the `Artifact` slice graph using a set of `go channels` to achieve the topologically sorted build order.
```

Comment:
Are we deprecating them?  You're planning to remove them, right?

Reply:
yes, deleting them

Reply:
fixed